[libc][POSIX][unistd] implement getsid

ZakyHermawan opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/126603

ZakyHermawan committed:
feat: implement getsid

Signed-off-by: ZakyHermawan <zaky.hermawan9615@gmail.com>

ZakyHermawan committed:
test: fix no member named 'libc_errno' on getsid_test

Signed-off-by: ZakyHermawan <zaky.hermawan9615@gmail.com>

michaelrj-google commented on libc/test/src/unistd/getsid_test.cpp:
```diff
@@ -0,0 +1,17 @@
+#include "src/unistd/getsid.h"
+#include "test/UnitTest/Test.h"
+
+TEST(LlvmLibcGetPidTest, GetCurrSID) {
+  pid_t sid = LIBC_NAMESPACE::getsid(0);
+  ASSERT_NE(sid, -1);
+  ASSERT_ERRNO_SUCCESS();
+}
```
For syscall wrapper tests we prefer to have one test where the syscall succeeds and one where the syscall fails. It looks like this is just checking the succeeds case so far.

michaelrj-google commented on libc/src/unistd/linux/CMakeLists.txt:
```diff
@@ -235,6 +235,21 @@ add_entrypoint_object(
+  DEPENDS
+    libc.hdr.types.pid_t
+    libc.hdr.fcntl_macros
+    libc.include.unistd
```
these includes need to be cleaned up. This function doesn't use `fcntl_macros` or `<unistd.h>`.

michaelrj-google reviewed (commented):
Thanks for this patch, it looks really good! There are some minor things to clean up, but after that this should be good to land. Thanks for contributing!

ZakyHermawan committed:
test: add test for getsid_test.cpp to check for failing getsid syscall
This commit also do cleanup on libc/src/unistd/linux/CMakeLists.txt to remove
unnecessary fnctl_macros and unistd from add_entrypoint_object for getsid

Signed-off-by: ZakyHermawan <zaky.hermawan9615@gmail.com>

ZakyHermawan committed:
Merge remote-tracking branch 'upstream/main' into implement-getsid

michaelrj-google commented on libc/test/src/unistd/getsid_test.cpp:
```diff
@@ -0,0 +1,21 @@
+  ASSERT_NE(sid, -1);
+  ASSERT_ERRNO_SUCCESS();
+
+  pid_t nonexist_sid = LIBC_NAMESPACE::getsid(999999);
```
I'd recommend using the sid `-1` since `999999` might theoretically be valid, which would cause flakey tests.

michaelrj-google reviewed (commented):
just one more small tweak and it should be ready to land

ZakyHermawan committed:
test: change test for failing getsid to use -1 as the argument
because calling getsid with -1 as the argument will guarantee the getsid syscall fails

Signed-off-by: ZakyHermawan <zaky.hermawan9615@gmail.com>

michaelrj-google commented:
LGTM, do you need me to merge this for you once the presubmits are done?

ZakyHermawan commented:
yes, please

ZakyHermawan committed:
Merge branch 'main' into implement-getsid

michaelrj-google commented:
Presubmit failures are unrelated, tests pass on my system. Merging.

jhuber6 commented:
Ignore that ^
